### PR TITLE
ft/Improve Posts

### PR DESCRIPTION
**Pull request by kelvinkipruto**
## Description
This PR implements some improvements on posts

1. Introduces a posts settings global to control whether to show recent posts on individual pages
2. Fixes a bug whereby you cannot sort posts by tag
3. Remove the featured post from the list of posts if it is available. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Screenshots
image

image

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by kilemensi — COMMENTED** [review]
👍🏽 

--

1. Is this block for [Projects](https://cfa.dev.codeforafrica.org/projects/civicsignal) or [Stories](https://cfa.dev.codeforafrica.org/stories/battle-for-gender-equality-in-african-media-continues)
2. If it's related stories, we'd need a way of relating. Once way is making sure they share tags. If it's recent, we need to fetch ordered by publication date descending or something.
3. How will the content editors add this block in the CMS? We don't have individual pages for projects or stories in the CMS.

File: apps/codeforafrica/src/payload/blocks/Posts.js [hunk 1]
```diff
@@ -73,6 +73,16 @@ const Posts = {
           type: "text",
           defaultValue: "Read More",
         },
+        {
```
Comment (koechkevin) [inline apps/codeforafrica/src/payload/blocks/Posts.js:76]: @kelvinkipruto  thinking out loud, is recent~Stories expected to be fixed by post? why not fetch from the collection ordering by publishedOn ?

File: apps/codeforafrica/src/lib/data/pagify/post.js [hunk 2]
```diff
@@ -1,10 +1,39 @@
...
+    api,
+    {
+      limit: 3,
+      sort: "publishedOn",
+    },
+    "stories",
+  );
+  const block = parentPage.blocks.find(
+    ({ blockType }) => blockType === "post-list",
+  );
```
Comment (koechkevin) [inline apps/codeforafrica/src/lib/data/pagify/post.js:18]: Let us check if block exists or else we will end up with undefined errors.

File: apps/codeforafrica/src/payload/blocks/Posts.js (hunk 1 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/payload/blocks/Posts.js:76]: @koechkevin I'm already fetching [ordered by publishedOn](https://github.com/CodeForAfrica/ui/blob/95c9b955435e94ab67696b83c8e6ede4c80cc393/apps/codeforafrica/src/lib/data/pagify/post.js#L8-L15). This is just the title for the Recent Stories

**Comment by kelvinkipruto** [comment]
@kilemensi @koechkevin This is now ready.

**Review by kilemensi — COMMENTED** [review]
My quick review on this @kelvinkipruto 

We should change website/settings to be:

 - Settings
   - Site
   - Publication
   - Users

**Site**: This is the current Settings global, just a rename i.e. All site-wide settings will be here.
**Publication**: This settings will cover everything under **Publication**. For now we can have just have Posts (as a tab) and under that, we can have Recent Stores, show recent stories checkbox and if checked, allow editors to set title for recent stories. Similar for Recent Opportunities.
**Users**: Remain as is.

**Comment by kilemensi** [comment]
Does my [comment](https://github.com/CodeForAfrica/ui/pull/621#pullrequestreview-1661709936) make sense @kelvinkipruto?

**Comment by kelvinkipruto** [comment]
@kilemensi Yes. Currently updating it.

File: apps/codeforafrica/src/payload/globals/Site/index.js [hunk 3]
```diff
@@ -2,8 +2,9 @@ import EngagementTab from "./EngagementTab";
 import GeneralTab from "./GeneralTab";
 import NavigationTab from "./NavigationTab";
 
-const Settings = {
-  slug: "settings",
+const Site = {
+  slug: "site",
```
Comment (kelvinkipruto) [inline apps/codeforafrica/src/payload/globals/Site/index.js:6]: @kilemensi Not sure about it, but since we already have some data in prod, do we need to update this slug, or should we keep it as it was: `settings`?

Comment (kilemensi) [reply apps/codeforafrica/src/payload/globals/Site/index.js:6]: Yes, let's update it @kelvinkipruto. I don't think the content team has had time to really dig in just yet. Keep a copy of the JSON data handy before deploying & we can update the data after deployment.

Comment (kilemensi) [reply apps/codeforafrica/src/payload/globals/Site/index.js:6]: For all the above changes, we can use slug as is i.e. `site`, `publication`, etc., or prepend `setting-` to show these are settings related i.e. `settings-site`, `settings-publication`, etc.

I'll let you be the architect.

File: apps/codeforafrica/src/components/Articles/Articles.js [hunk 4]
```diff
@@ -35,7 +35,7 @@ const Articles = React.forwardRef(function Articles(props, ref) {
   const [q, setQ] = useState();
   const [filtering, setFiltering] = useState(false);
   const [tag, setTag] = useState(allTag);
-  const queryParams = useFilterQuery({ page, q, tag: tag.slug });
+  const queryParams = useFilterQuery({ page, q, tag });
```
Comment (kilemensi) [inline apps/codeforafrica/src/components/Articles/Articles.js:38]: Why the change from `tag.slug` to just `tag`?

File: apps/codeforafrica/src/components/Articles/useArticles.js [hunk 5]
```diff
@@ -4,11 +4,13 @@ import useFilterQuery from "@/codeforafrica/components/useFilterQuery";
...
-function useArticles(params, primaryTag) {
+function useArticles(params, { primaryTag, featured }) {
   const queryParams = useFilterQuery(params);
   const query = queryParams
-    ? `${queryParams}&primaryTag=${primaryTag}`
-    : `?primaryTag=${primaryTag}`;
+    ? `${queryParams}&primaryTag=${primaryTag}${
+        featured ? `&featured=${featured}` : ""
+      }`
+    : `?primaryTag=${primaryTag}${featured ? `&featured=${featured}` : ""}`;
```
Comment (kilemensi) [inline apps/codeforafrica/src/components/Articles/useArticles.js:13]: This can be improved e.g.
```js
  const filterQuery = queryParams ? `?${queryParams}&` : '?';
  const primaryTagQuery = `primaryTag=${primaryTag}`;
  const featuredQuery = featured ? `&featured=${featured}&` : '';
  const query = `${filterQuery}${primaryTagQuery}${featuredQuery}`
```

File: apps/codeforafrica/src/lib/data/blockify/posts.js [hunk 6]
```diff
@@ -11,6 +11,13 @@ async function posts(block, api, context) {
   const featured = featuredStory ? formatPost(featuredStory, primaryTag) : null;
   const options = {
     ...query,
+    ...(featured && {
+      where: {
+        slug: {
+          not_equals: featured.slug,
```
Comment (kilemensi) [inline apps/codeforafrica/src/lib/data/blockify/posts.js:17]: Let's make sure the `slug` field has [`index: true`](https://payloadcms.com/docs/fields/text#config) if we're going to be using it in queries.

File: apps/codeforafrica/src/payload/globals/Publication/PostsTab.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
+const PostsTab = {
+  label: "Posts",
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/globals/Publication/PostsTab.js:2]:
```suggestion
  label: "Post",
```

Since this is for individual posts, may be we should name it so.

File: apps/codeforafrica/src/payload/globals/Publication/PostsTab.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+          label: "Primary Tag",
+          type: "text",
+          required: true,
+          defaultValue: "stories",
+          admin: {
+            hidden: true,
+          },
+        },
+      ],
+    },
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/globals/Publication/PostsTab.js:40]: 1. UI needs to look like the rest of the globals
2. localisation must be enabled at top levels similar to other globals
```suggestion
    {
      type: "collapsible",
      label: "Story",
      fields: [
        {
          name: "recent-stories",
          label: "Recent stories",
          type: "group",
          localized: true,
          fields: [
            {
              name: "enabled",
              label: "Show recent stories",
              type: "checkbox",
              required: true,
              defaultValue: false,
            },
            {
              name: "title",
              label: "Title",
              type: "text",
              required: true,
              defaultValue: "Recent Stories",
              admin: {
                condition: (_, data) => data.showRecent,
              },
            },
            {
              name: "primaryTag",
              label: "Primary Tag",
              type: "text",
              required: true,
              defaultValue: "stories",
              admin: {
                hidden: true,
              },
            },
          ],
          admin: {
            className: "group-field-nested",
          },
        },
      ],
    },
```
Screenshot 2023-10-09 at 09 49 58

File: apps/codeforafrica/src/payload/globals/Publication/index.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import PostsTab from "./PostsTab";
+
+const Publication = {
+  slug: "publication",
+  label: "Publication",
+  access: {
+    read: () => true,
+  },
+  admin: {
+    group: "Website",
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/globals/Publication/index.js:10]:
```suggestion
    group: "Settings",
```

`admin.group` in `Users` and `Site` needs to be updated as well.

**Review by koechkevin — COMMENTED** [review]
@kelvinkipruto are we planning to add migrations since we are changing slug and production data already exists?

**Comment by kelvinkipruto** [comment]
> are we planning to add migrations since we are changing slug and production data already exists?

@koechkevin I have backed up the JSON data currently in use. We will use it when we merge this.

File: apps/codeforafrica/src/components/Articles/Articles.js (hunk 4 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/components/Articles/Articles.js:38]: @kilemensi #610 [updated the `useFilterQuery`](https://github.com/CodeForAfrica/ui/blob/e3eb9f96e118f6cdf0dee21d7764afb0c8270b6b/apps/codeforafrica/src/components/useFilterQuery/useFilterQuery.js#L17-L19) hook to take the whole tag, rather than just the slug. That is why currently in production you can't filter stories & opportunities by Tag.

File: apps/codeforafrica/src/components/Articles/useArticles.js [hunk 10]
```diff
@@ -4,11 +4,13 @@ import useFilterQuery from "@/codeforafrica/components/useFilterQuery";
...
 
-function useArticles(params, primaryTag) {
+function useArticles(params, { primaryTag, featured }) {
   const queryParams = useFilterQuery(params);
-  const query = queryParams
-    ? `${queryParams}&primaryTag=${primaryTag}`
-    : `?primaryTag=${primaryTag}`;
+  const separator = queryParams ? "" : "?";
+  const filterQuery = queryParams ? `${queryParams}&` : "";
+  const primaryTagQuery = `primaryTag=${primaryTag}`;
```
Comment (kilemensi) [inline apps/codeforafrica/src/components/Articles/useArticles.js:11]: I think this is not right-ish. `separator` should always be `?` because there will always be `primaryTagQuery` even where there isn't `queryParams`.

File: apps/codeforafrica/src/lib/data/utils/posts.js [hunk 11]
```diff
@@ -117,29 +118,54 @@ export async function getPost(api, slug, primaryTag) {
...
+      }),
+      primaryTag,
+      blockType: "article",
+      ...other,
+    },
+  ];
+
+  const globalPostSettings = await api.findGlobal("settings-publication");
+  const primaryTagGlobalSettings = globalPostSettings?.[primaryTag] ?? {};
+  const { showRecent, title: recentTitle } = primaryTagGlobalSettings;
```
Comment (kilemensi) [inline apps/codeforafrica/src/lib/data/utils/posts.js:145]:
```suggestion
  const publicationSettings = await api.findGlobal("settings-publication");
  const primaryTagPostSettings = publicationSettings?.[primaryTag] ?? {};
  const { showRecent, title: recentTitle } = primaryTagPostSettings;
```

No need to use global.

File: apps/codeforafrica/src/pages/api/v1/posts.page.js [hunk 12]
```diff
@@ -2,12 +2,25 @@ import { getPosts } from "@/codeforafrica/lib/data/utils/posts";
...
   try {
-    const data = await getPosts(api, other, primaryTag);
+    const data = await getPosts(
+      api,
+      {
+        ...other,
+        ...(featured && {
+          where: {
+            slug: {
+              not_equals: featured,
```
Comment (kilemensi) [inline apps/codeforafrica/src/pages/api/v1/posts.page.js:17]: Does this work well even when `featured` is `undefined`?

File: apps/codeforafrica/src/payload/collections/Posts.js [hunk 13]
```diff
@@ -39,7 +40,9 @@ const Posts = {
       },
     }),
     content({ minRows: 1, required: true, localized: true }),
-    slug(),
+    slug({
+      index: true,
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/collections/Posts.js:44]: Add this to `payload/fields/slug.js`. Should be beneficial to everything that's using a `slug`.

File: apps/codeforafrica/src/payload/collections/Posts.js [hunk 14]
```diff
@@ -23,6 +23,7 @@ const Posts = {
     description: "Stories and Opportunities",
     group: "Publication",
     useAsTitle: "title",
+    listSearchableFields: ["content", "excerpt"],
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/collections/Posts.js:26]: No need to include `title`?

File: apps/codeforafrica/src/payload/globals/Publication/PostTab.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            {
+              name: "primaryTag",
+              label: "Primary Tag",
+              type: "text",
+              required: true,
+              defaultValue: "stories",
+              admin: {
+                hidden: true,
+              },
+            },
```
Comment (kilemensi) [inline apps/codeforafrica/src/payload/globals/Publication/PostTab.js:40]: Thinking out loud, do we need this field if the group is already named "stories" or "opportunities" i.e. `group.name` matches `primaryTag`.

**Review by kilemensi — APPROVED** [review]

File: apps/codeforafrica/src/payload/globals/Publication/PostTab.js (hunk 15 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/payload/globals/Publication/PostTab.js:40]: I agree. We should remove this as it is not used anywhere.

File: apps/codeforafrica/src/payload/collections/Posts.js (hunk 14 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/payload/collections/Posts.js:26]: @kilemensi You can already search by title. I think it is made possible by `useAsTitle: "title"`

File: apps/codeforafrica/src/payload/collections/Posts.js (hunk 13 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/payload/collections/Posts.js:44]: Turns out this is already indexed

Comment (kelvinkipruto) [reply apps/codeforafrica/src/payload/collections/Posts.js:44]: Turns out this is already indexed

File: apps/codeforafrica/src/pages/api/v1/posts.page.js (hunk 12 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/pages/api/v1/posts.page.js:17]: @kilemensi I Just tested it out and it works. Also, we are making sure to add `featured` only if a featured article exists.

File: apps/codeforafrica/src/components/Articles/useArticles.js (hunk 10 above)
Comment (kelvinkipruto) [reply apps/codeforafrica/src/components/Articles/useArticles.js:11]: @kilemensi [`useFilterQuery` already](https://github.com/CodeForAfrica/ui/blob/73018e2697cd8ed2f84fc73f786c27f1a961ec56/apps/codeforafrica/src/components/useFilterQuery/useFilterQuery.js#L21) appends `?` if there are `queryParams`. If `queryParams` does not exist, then we have to use create `?` for the `primaryTagQuery`

Comment (kilemensi) [reply apps/codeforafrica/src/components/Articles/useArticles.js:11]: If it's there, then all good @kelvinkipruto
